File: docs/demos/structs/deque.c
// -*- coding: utf-8 -*-

#include <stdio.h>
#include <stdint.h>
#include <string.h>


typedef struct Deque{
  void * contents;
  int max_size;
  int itemsize;
  int head_index;
  int tail_index;
} Deque;


int append(Deque * self, void * item) {
  int index = self -> head_index;
  memcpy(self -> contents + index * self -> itemsize, item, self -> itemsize);
  self -> head_index += 1;
  self -> head_index %= self -> max_size;
  return index;
}


int prepend(Deque * self, void * item) {
  int index = self -> tail_index;
  memcpy(self -> contents + index * self -> itemsize, item, self -> itemsize);
  self -> tail_index -= 1;
  if (self -> tail_index < 0) self -> head_index += self -> max_size;
  return index;
}

File: docs/demos/structs/deque.py
import numpy as np
from cslug import CSlug, anchor, ptr

slug = CSlug(anchor("deque.c"))
slug.make()


class Deque(object):

    def __init__(self, max_size, dtype):
        self.dtype = np.dtype(dtype)
        self.max_size = max_size
        self._contents = np.empty(max_size, self.dtype)

        self._raw = slug.dll.Deque(ptr(self._contents), self.max_size,
                                   self.dtype.itemsize, 0, self.max_size - 1)


if __name__ == "__main__":
    self = Deque(10, np.int)
    item = np.asarray(12)
    assert slug.dll.prepend(self._raw._ptr, ptr(item)) == 9
    assert self._contents[9] == 12
